src/meepgeom.cpp

```diff
@@ -2521,63 +2583,47 @@ void get_material_tensor(const medium_struct *mm, double freq,
+*/
 double get_material_gradient(
-    double u,                       // material parameter at current point
-    std::complex<double> fields_a,  // adjoint field at current point
-    std::complex<double> fields_f,  // forward field at current point
-    double freq,                    // frequency
-    material_data *md,              // material
-    meep::component field_dir,      // current field component
-    double du                       // step size
+    const meep::vec &r,              // current postion
+    std::complex<double> *fields_a,  // adjoint field at current point
+    std::complex<double> *fields_f,  // forward field at current point
+    double freq,                     // frequency
+    geom_epsilon *geps,              // geometry epsilon object
+    double du                        // step size
 ) {
-  /* Note that the current implementation assumes that
-  no materials have off-diag components and that if a material
-  has dispersion, it is either a lorentzian or drude profile.
-  */
-
-  const medium_struct *mm = &(md->medium);
-  const medium_struct *m1 = &(md->medium_1);
-  const medium_struct *m2 = &(md->medium_2);
-
-  // trivial case
-  if ((mm->E_susceptibilities.size() == 0) && mm->D_conductivity_diag.x == 0 &&
-      mm->D_conductivity_diag.y == 0 && mm->D_conductivity_diag.z == 0){
-        switch (field_dir){
-          case meep::Er:
-          case meep::Ex: return (m2->epsilon_diag.x - m1->epsilon_diag.x) * (fields_a * fields_f).real();
-          case meep::Ep:
-          case meep::Ey: return (m2->epsilon_diag.y - m1->epsilon_diag.y) * (fields_a * fields_f).real();
-          case meep::Ez: return (m2->epsilon_diag.z - m1->epsilon_diag.z) * (fields_a * fields_f).real();
-          default: meep::abort("Invalid field component specified in gradient.");
-        }
-      }
-
   /* For now we do a finite difference approach to estimate the
   gradient of the system matrix `A` since it's fairly accurate,
   cheap, and easy to generalize. */
-  std::complex<double> dA_du_0[9] = {std::complex<double>(0, 0)};
-  epsilon_material_grid(md, u - du);
-  get_material_tensor(mm, freq, dA_du_0);
-
-  std::complex<double> dA_du_1[9] = {std::complex<double>(0, 0)};
-  epsilon_material_grid(md, u + du);
-  get_material_tensor(mm, freq, dA_du_1);
-
-  std::complex<double> dA_du[9] = {std::complex<double>(0, 0)};
-  for (int i = 0; i < 9; i++)
-    dA_du[i] = (dA_du_1[i] - dA_du_0[i]) / (2 * du);
-
-  int dir_idx = 0;
-  if (field_dir == meep::Ex || field_dir == meep::Er)
-    dir_idx = 0;
-  else if (field_dir == meep::Ey || field_dir == meep::Ep)
-    dir_idx = 1;
-  else if (field_dir == meep::Ez)
-    dir_idx = 2;
-  else
-    meep::abort("Invalid adjoint field component");
-
-  std::complex<double> result = fields_a * dA_du[3 * dir_idx + dir_idx] * fields_f;
+  symm_matrix chi1_1, chi1_2, dA_du;
+
+  geps->u_p = du;                   // perturbation term in one direction
+  get_chi1_tensor(r, geps, &chi1_1);
```

Comment:
There is a bit of a complication here that is being overlooked. At the Yee grid point for e.g. *E<sub>x</sub>* (where the forward and adjoint fields for this component are computed by Meep), the row from the tensor ∂ε/∂ρ containing the effective ε (in this case containing the elements [ε<sub>xx</sub>, ε<sub>xy</sub>, ε<sub>xz</sub>]) is multiplied by the column vector of the adjoint fields [*E<sup>a</sup><sub>x</sub>*, *E<sup>a</sup><sub>y</sub>*, *E<sup>a</sup><sub>z</sub>*] to obtain "one side of the inner product." The complication, of course, is that *E<sub>y</sub>* and *E<sub>z</sub>* are not defined at the Yee grid point for *E<sub>x</sub>* which means they have to be interpolated to that point using the nearest grid values.

In the way that it is currently set up, there does not seem to be any such interpolation which means that an error is being introduced into the gradient at this step of the computation.